packages/neon-core/src/tx/transaction/Transaction.ts

```diff
@@ -1,104 +1,386 @@
+    }
+    this.scripts.push(witness);
+    this.scripts = this.scripts.sort(
+      (w1, w2) => parseInt(w1.scriptHash, 16) - parseInt(w2.scriptHash, 16)
+    );
+    return this;
   }
 
-  public get [Symbol.toStringTag](): string {
-    return "Transaction";
+  /**
+   * Adds some script intents to the Transaction
+   * System Fee will be increased automatically
+   * However system Fee calculated in this method is insufficient if calling non-native contracts.
+   * If the transaction is invoking contracts other than native contracts, invokeScript rpc request can test the systemFee.
+   * @param scriptIntents sciprt Intents to add to the transaction
+   */
+  public addIntent(...scriptIntents: ScriptIntent[]): this {
+    let increasedSystemFee = new Fixed8(0);
+    this.script = scriptIntents.reduce((accumulatedScript, intent) => {
+      const { hex, fee } = createScript(intent);
+      this._pre_systemFee = this._pre_systemFee.plus(fee);
+      increasedSystemFee = increasedSystemFee.plus(fee);
+      this.intents.push(intent);
+      return accumulatedScript + hex;
+    }, this.script);
+    log.info(
+      `Increased systemFee: ${increasedSystemFee.toNumber()}, totally ${this.systemFee.toNumber()}`
+    );
+    return this;
   }
 
   /**
-   * Exclusive Data
+   * If transaction script is constructed only by method `addIntent`, systemFee will be calcuated.
+   * If all intents are about native contract invocation, this system fee is accurate.
+   * In this case, you can use this method to get minimum system fee.
    */
-  public get exclusiveData(): { [key: string]: any } {
-    throw new Error("Not Implemented!");
+  public useCalculatedSystemFee(): void {
+    this.systemFee = this._pre_systemFee;
   }
 
-  public get fees(): number {
-    return 0;
+  /**
+   * This function regulate system fee with 2 processes:
+   * 1. considering the free system fee threshold
+   * 2. ceil systemFee with factor
+   */
+  public regulateSystemFee(): void {
+    let systemFee = this.systemFee.minus(SYSTEM_FEE_FREE);
+    systemFee = systemFee.comparedTo(0) >= 0 ? systemFee : new Fixed8(0);
+    if (systemFee.comparedTo(0) > 0) {
+      const remainder = systemFee.mod(SYSTEM_FEE_FACTOR);
+      if (remainder.comparedTo(0) > 0) {
+        this.systemFee = systemFee.plus(
+          new Fixed8(SYSTEM_FEE_FACTOR).minus(remainder)
+        );
+      } else {
+        this.systemFee = systemFee;
+      }
+    } else {
+      this.systemFee = new Fixed8(0);
+    }
+  }
+
+  /**
+   * calculate networkFee
+   * networkFee = verificationCost + txSize * POLICY_FEE_PERBYTE
+   * @param autoAjust when true, auto justify tx networkfee to the minimum value
+   * @returns minimum networkFee that this tx needs
+   */
+  public calculateNetworkFee(autoAdjust: boolean = true): Fixed8 {
+    const signers = this.getScriptHashesForVerifying();
+    // compute the size of transaction.
+    const txHeaderSize =
+      /* version */ 1 +
+      /* nonce */ 4 /* sender */ +
```

Comment:
is the nonce always a fixed size?

Reply:
Yes. I think so.
```c#
public const int HeaderSize =
            sizeof(byte) +  //Version
            sizeof(uint) +  //Nonce
```
Above is part of the code to calculate tx header size.